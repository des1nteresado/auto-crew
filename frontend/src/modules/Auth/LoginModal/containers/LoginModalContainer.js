import React, { useState, useCallback, useMemo, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';

import Snack from 'common/components/SnackBar';
import { validateEmail, validatePassword } from 'helpers/validations';
import LoginModal from '../components/LoginModal';
import { CLEAR_ERRORS, LOGIN } from '../actions';

const initialState = { email: '', password: '' };

const LoginModalContainer = ({
  isLoginModalShowed,
  handleShowLoginModal,
  handleShowResetPasswordModal,
  handleShowRegisterModal,
}) => {
  const [formData, setFormData] = useState(initialState);
  const [isLoginError, setIsLoginError] = useState(false);

  const dispatch = useDispatch();

  const { errors } = useSelector((state) => state.user);

  useEffect(() => setIsLoginError(false), []);

  useEffect(() => {
    setFormData(initialState);
  }, [isLoginModalShowed]);

  const isEmailValid = useMemo(() => validateEmail(formData.email), [formData.email]);
  const isPasswordValid = useMemo(() => validatePassword(formData.password), [formData.password]);

  const showRegisterModalHandler = useCallback(() => {
    handleShowLoginModal(false);
    handleShowRegisterModal(true);
  }, [handleShowLoginModal, handleShowRegisterModal]);

  const showResetPasswordModalHandler = useCallback(() => {
    handleShowLoginModal(false);
    handleShowResetPasswordModal(true);
  }, [handleShowLoginModal, handleShowResetPasswordModal]);

  const onChangeFields = useCallback(
    (event) => {
      setFormData({
        ...formData,
        [event.target.name]: event.target.value,
      });
      setIsLoginError(false);
    },
    [formData, setFormData, setIsLoginError]
  );

  const handleSignIn = useCallback(
    (event) => {
      event.preventDefault();

      dispatch(LOGIN(formData));
    },
    [formData, dispatch]
  );

  const handleErrorClear = useCallback(() => {
    dispatch(CLEAR_ERRORS());
  }, [dispatch]);

  return (
    <>
      <LoginModal
        formData={formData}
        onChange={onChangeFields}
        onSubmit={handleSignIn}
        isEmailValid={isEmailValid}
        isPasswordValid={isPasswordValid}
        isLoginError={isLoginError}
        isLoginModalShowed={isLoginModalShowed}
        handleShowLoginModal={handleShowLoginModal}
        handleShowResetPasswordModal={showResetPasswordModalHandler}
        handleShowRegisterModal={showRegisterModalHandler}
      />
      <Snack message={errors} onHandleClose={handleErrorClear} />
    </>
  );
};

LoginModalContainer.propTypes = {
  isLoginModalShowed: PropTypes.bool.isRequired,
  handleShowLoginModal: PropTypes.func.isRequired,
  handleShowResetPasswordModal: PropTypes.func.isRequired,
  handleShowRegisterModal: PropTypes.func.isRequired,
};

export default React.memo(LoginModalContainer);
