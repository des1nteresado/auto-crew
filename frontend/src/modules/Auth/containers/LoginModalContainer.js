import React, { useState, useCallback, useMemo, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';

import Snack from 'common/components/SnackBar';
import { validateEmail, validatePassword } from 'helpers/validations';
import LoginModal from '../components/LoginModal';
import { CLEAR_ERRORS, SIGN_IN } from '../actions';
import { useCommonContext } from 'common/contexts/CommonContext';
import { userSelector } from '../selectors';

const initialState = { email: '', password: '' };

const LoginModalContainer = () => {
  const {
    isLoginModalShowed,
    handleShowLoginModal,
    handleShowResetPasswordModal,
    handleShowRegisterModal,
  } = useCommonContext();

  const [formData, setFormData] = useState(initialState);
  const [isLoginError, setIsLoginError] = useState(false);

  const dispatch = useDispatch();

  const { message, isSuccess } = useSelector(userSelector);

  useEffect(() => setIsLoginError(false), []);

  useEffect(() => {
    setFormData(initialState);
  }, [isLoginModalShowed]);

  useEffect(() => {
    handleShowLoginModal(false);
  }, [isSuccess]);

  const isEmailValid = useMemo(() => validateEmail(formData.email), [formData.email]);
  const isPasswordValid = useMemo(() => validatePassword(formData.password), [formData.password]);

  const showRegisterModalHandler = useCallback(() => {
    handleShowLoginModal(false);
    handleShowRegisterModal(true);
  }, [handleShowLoginModal, handleShowRegisterModal]);

  const showResetPasswordModalHandler = useCallback(() => {
    handleShowLoginModal(false);
    handleShowResetPasswordModal(true);
  }, [handleShowLoginModal, handleShowResetPasswordModal]);

  const onChangeFields = useCallback(
    (event) => {
      setFormData({
        ...formData,
        [event.target.name]: event.target.value,
      });
      setIsLoginError(false);
    },
    [formData, setFormData, setIsLoginError]
  );

  const handleSignIn = useCallback(
    (event) => {
      event.preventDefault();

      dispatch(SIGN_IN(formData));
    },
    [formData, dispatch]
  );

  const handleErrorClear = useCallback(() => {
    dispatch(CLEAR_ERRORS());
  }, [dispatch]);

  return (
    <>
      <LoginModal
        formData={formData}
        onChange={onChangeFields}
        onSubmit={handleSignIn}
        isEmailValid={isEmailValid}
        isPasswordValid={isPasswordValid}
        isLoginError={isLoginError}
        isLoginModalShowed={isLoginModalShowed}
        handleShowLoginModal={handleShowLoginModal}
        handleShowResetPasswordModal={showResetPasswordModalHandler}
        handleShowRegisterModal={showRegisterModalHandler}
      />
      <Snack message={message} onHandleClose={handleErrorClear} />
    </>
  );
};

export default LoginModalContainer;
