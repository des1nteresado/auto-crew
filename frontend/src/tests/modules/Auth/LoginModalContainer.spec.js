import React from 'react';
import { mount } from 'enzyme';

import LoginModalContainer from 'modules/Auth/LoginModal/containers/LoginModalContainer';

const defaultProps = {
  isLoginModalShowed: true,
  handleShowLoginModal: jest.fn(),
  handleShowResetPasswordModal: jest.fn(),
  handleShowRegisterModal: jest.fn(),
};

const setUp = (props) => mount(<LoginModalContainer {...defaultProps} {...props} />).dive();

describe('LoginModalContainer', () => {
  let component;

  beforeEach(() => {
    component = setUp();
  });

  it('renders LoginModalContainer component correctly', () => {
    expect(component).toMatchSnapshot();
  });
});
