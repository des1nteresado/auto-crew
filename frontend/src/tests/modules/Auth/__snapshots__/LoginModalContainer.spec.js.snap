// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginModalContainer renders LoginModalContainer component correctly 1`] = `undefined`;
